proposals/list-patterns.md

```diff
@@ -121,6 +172,141 @@ expr.Length is >= 2
 ```
 The *input type* for the *slice_pattern* is the return type of the underlying `this[Range]` or `Slice` method with two exceptions: For `string` and arrays, `string.Substring` and `RuntimeHelpers.GetSubArray` will be used, respectively.
 
+#### Lowering on enumerable type
+
+> **Open question**: Confirm that async enumerables are out-of-scope.  
+> **Open question**: Confirm that slice patterns with a sub_pattern (such as `..var x`) are out-of-scope.  
+
+Although a helper type is not necessary, it helps simplify and illustrate the logic.
+
+```
+class ListPatternHelper
+{
+  // Notes: 
+  // We could inline this logic to avoid creating a new type and to handle the pattern-based enumeration scenarios.
+  // We may only need one element in start buffer, or maybe none at all, if we can control the order of checks in the patterns DAG.
+  // We could emit a count check for a non-terminal `..` and economize on count checks a bit.
+  private EnumeratorType enumerator;
+  private int count;
+  private ElementType[] startBuffer;
+  private ElementType[] endCircularBuffer;
+
+  public ListPatternHelper(EnumerableType enumerable, int startPatternsCount, int endPatternsCount)
+  {
+    count = 0;
+    enumerator = enumerable.GetEnumerator();
+    startBuffer = startPatternsCount == 0 ? null : new ElementType[startPatternsCount];
```

Comment:
And having an `IDisposable` implementation for returning them.

Reply:
Right. Since we'll need to track disposal of enumerator we have some options here.
Stackalloc might be an option too if we inline this logic (then the state is just locals to the expression).

---
In reply to: [601740384](https://github.com/dotnet/csharplang/pull/4575#discussion_r601740384) [](ancestors = 601740384)

Reply:
I don't think we'll want to use stackalloc (unconditionally, anyway). This might be a place where we'd like a runtime helper like https://github.com/dotnet/runtime/issues/25423.